# This script includes the following points:

Step 3. Importing the libraries and data

Step 3.1. Exploring the data


Step 4. Data wrangling

Step 4.1. Renaming columns

Step 4.2. Checking variable types

Step 5. Data quality and consistency

Step 5.1. Checking for mixed data

Step 5.2. Checking for missing values

Step 5.3. Dropping columns

Step 5.4. Checking for duplicates

Step 6. Combining data

Step 6.1 Merging dataframes

Step 7. Exporting data in pkl format.

# Step 3. Importing libraries and data

In [1]:
#Import libraries 
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/buketoztekin/Documents/Instacart Basket Analysis/'

In [25]:
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)


## Step 3.1 Exploring data

In [26]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [27]:
df_cust.shape

(206209, 10)

In [28]:
df_cust.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [33]:
medians = df_cust[['Age', 'income', 'n_dependants']].median()

In [34]:
print(medians)

Age                49.0
income          93547.0
n_dependants        1.0
dtype: float64


Based on the data exploration, we observe the following key points:

Count Consistency: The counts for all variables are identical, indicating there are no missing values in the dataset.

Age: The ages range from a minimum of 18 to a maximum of 81.
The mean and median ages are equal, suggesting that the age distribution is approximately normal.

Income: Similar to age, the mean and median incomes are the same, indicating that the income data is also normally distributed.

Number of Dependents: The number of dependents varies from a minimum of 1 to a maximum of 3.
The values for the number of dependents appear to be normally distributed within this range.

# Step 4. Data Wrangling

In [37]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## Step 4.1. Renaming columns

In [38]:
#Renaming First Name 
df_cust.rename(columns = {'First Name' : 'first_name'}, inplace = True)
df_cust.rename(columns = {'Gender' : 'gender'}, inplace = True)
df_cust.rename(columns = {'Surnam' : 'surname'}, inplace = True)
df_cust.rename(columns = {'STATE' : 'state'}, inplace = True)
df_cust.rename(columns = {'Age' : 'age'}, inplace = True)
df_cust.rename(columns = {'First Name' : 'first_name'}, inplace = True)
df_cust.rename(columns = {'n_dependants' : 'number_of_dependants'}, inplace = True)
df_cust.rename(columns = {'fam_status' : 'family_status'}, inplace = True)


In [39]:
df_cust.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

## Step 4.2 Checking variable types

In [ ]:
df_cust.dtypes

# Step 5. Data Quality and Consistency Checks

## Step 5.1 Checking for Mixed Data

In [40]:
# Checking for mixed data 
for col in df_cust.columns.tolist():
    if df_cust[col].apply(type).nunique() > 1:
        print(col)

first_name


In [41]:
#Fixing mixed data in 'first_name' column
df_cust['first_name'] = df_cust['first_name'].astype('str')


In [42]:
# Checking for mixed data after fixing
for col in df_cust.columns.tolist():
    if df_cust[col].apply(type).nunique() > 1:
        print(col)

## Step 5.2 Checking for Missing Values

In [43]:
#Finding missing values
df_cust.isnull().sum()


user_id                 0
first_name              0
surname                 0
gender                  0
state                   0
age                     0
date_joined             0
number_of_dependants    0
family_status           0
income                  0
dtype: int64

It seems like there are no missing values. But when we skim through the dataframe, some 'nan' values can be detected. These values are possibly detected as strings rather than NaN values. 

In [44]:
df_cust.head(60)

user_id   first_name    surname  gender           state  age date_joined  \
0     26711      Deborah   Esquivel  Female        Missouri   48    1/1/2017   
1     33890     Patricia       Hart  Female      New Mexico   36    1/1/2017   
2     65803      Kenneth     Farley    Male           Idaho   35    1/1/2017   
3    125935     Michelle      Hicks  Female            Iowa   40    1/1/2017   
4    130797          Ann    Gilmore  Female        Maryland   26    1/1/2017   
5    133128      Cynthia      Noble  Female        Kentucky   43    1/1/2017   
6    152052        Chris     Walton    Male         Montana   20    1/1/2017   
7    168851       Joseph    Hickman    Male  South Carolina   30    1/1/2017   
8     69965       Jeremy       Vang    Male           Texas   47    1/1/2017   
9     82820        Shawn      Chung    Male        Virginia   26    1/1/2017   
10    26441       Gloria   Stafford  Female          Nevada   19    1/1/2017   
11    33835        Roger      Casey    Male        Nebraska   45    1/1/2017   
12    75032      Cynthia   Mcdaniel  Female         Georgia   75    1/1/2017   
13       51    Stephanie     Jensen  Female         Wyoming   37    1/1/2017   
14      516        Peter       Hunt    Male        Colorado   51    1/1/2017   
15     6965      Brandon       Mata    Male    North Dakota   70    1/1/2017   
16     8794          Amy       Mack  Female        Nebraska   30    1/1/2017   
17     9944        Brian     Oneill    Male       Wisconsin   26    1/1/2017   
18    13313       Victor       Moss    Male          Alaska   31    1/1/2017   
19    17335      Patrick     Fowler    Male         Vermont   65    1/1/2017   
20    29864        Randy      Mills    Male    North Dakota   39    1/1/2017   
21    37744        Wanda      Salas  Female        Arkansas   77    1/1/2017   
22    50000       Sandra      Trejo  Female           Maine   64    1/1/2017   
23    55567       Martha      Hurst  Female  North Carolina   67    1/1/2017   
24    57549        Shawn      Moore    Male         Montana   72    1/1/2017   
25    60386        Wayne   Woodward    Male          Alaska   59    1/1/2017   
26    68032      Charles  Stevenson    Male   West Virginia   74    1/1/2017   
27    83910         Ruth       Rice  Female         Indiana   70    1/1/2017   
28    86936     Kimberly     Monroe  Female          Oregon   57    1/1/2017   
29    99199     Kimberly     Monroe  Female        Arkansas   80    1/1/2017   
30   100123        Steve     Rogers    Male         Florida   31    1/1/2017   
31   104359         Carl     Benton    Male           Idaho   66    1/1/2017   
32   112664     Nicholas      Logan    Male      California   64    1/1/2017   
33   113616      Stephen    Murillo    Male    Pennsylvania   57    1/1/2017   
34   117740         Lisa     Sparks  Female          Oregon   55    1/1/2017   
35   127020         Judy    Barrett  Female            Ohio   35    1/1/2017   
36   135871        Doris      Mckee  Female     Connecticut   80    1/1/2017   
37   139692      Barbara      Chang  Female         Arizona   54    1/1/2017   
38   149755         Adam     Brewer    Male       Louisiana   72    1/1/2017   
39   150670        Terry  Wilkerson    Male            Iowa   55    1/1/2017   
40   155598        Marie      Beard  Female      Washington   73    1/1/2017   
41   159323      Rebecca       Tran  Female       Wisconsin   75    1/1/2017   
42   165553      Lillian   Schaefer  Female     Connecticut   71    1/1/2017   
43   166899      Kathryn   Guerrero  Female        New York   56    1/1/2017   
44   168047      Shirley   Harrison  Female          Alaska   37    1/1/2017   
45   180190      Douglas     Hebert    Male     Connecticut   30    1/1/2017   
46   180333  Christopher      Munoz    Male      Washington   80    1/1/2017   
47   195665      Brandon   Mckinney    Male    North Dakota   57    1/1/2017   
48   197419        Craig      Clark    Male   West Virginia   36    1/1/2017   
49  

In [45]:
#Replacing 'nan' values (as strings) with actual NaN
df_cust = df_cust.replace('nan', np.nan)

In [46]:
#Finding missing values
df_cust.isnull().sum()

user_id                     0
first_name              11259
surname                     0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
family_status               0
income                      0
dtype: int64

11259 first_name values are missing. Since we have the unique user_id for each individual, first_name and surnames columns can be dropped. If it is needed to know the names of the customers later, it is possible to find the through their user_ids.

## Step 5.3. Dropping Columns

In [47]:
#Dropping first_name and surname columns
df_cust_drop = df_cust.drop(columns = ['first_name', 'surname'])


In [48]:
df_cust_drop.head()

user_id  gender       state  age date_joined  number_of_dependants  \
0    26711  Female    Missouri   48    1/1/2017                     3   
1    33890  Female  New Mexico   36    1/1/2017                     0   
2    65803    Male       Idaho   35    1/1/2017                     2   
3   125935  Female        Iowa   40    1/1/2017                     0   
4   130797  Female    Maryland   26    1/1/2017                     1   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

## Step 5.4 Checking for Duplicates

In [49]:
#Checking for duplicates
df_dups = df_cust_drop[df_cust_drop.duplicated()]

In [50]:
df_dups.shape

(0, 8)

No duplicates were found.

# Step 6. Combining Data

In [3]:
# Importing prepared data
file_path = os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_merged_derived.pkl')
df_ords_prods_merge = pd.read_pickle(file_path)

In [52]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most orders   
1  Mid-range product  Regularly busy  Regularly busy        Average orders   
2  Mid-range product     Busiest day    Busiest days        Average orders   
3  Mid-range product  Regularly busy    Slowest days           Most orders   
4  Mid-range product      Least busy    Slowest days        Average orders   

  max_order      loyalty_flag average_spending spender_flag  \
0        32  Regular customer         6.935811  Low spender   
1        32  Regular customer         6.935811  Low spender   
2         5      New customer         7.930208  Low spender   
3         3      New customer         4.972414  Low spender   
4         3      New customer         4.972414  Low spender   

  purchase_frequency     frequency_flag  
0                8.0  Frequent customer  
1                8.0  Frequent customer  
2                7.0  Frequent customer  
3                9.0  Frequent customer  
4                9.0  Frequent customer  

[5 rows x 25 columns]

Since both df_cust_drop and the merged data has user_id column in common, these tables will be merged on user_id column.


In [53]:
df_ords_prods_merge.shape

(32404859, 25)

## Step. 6.1. Merging dataframes

In [54]:
#Merging df_cust_drop and df_ords_prods_merge on user_id
df_ords_prods_cust_merged = df_cust_drop.merge(df_ords_prods_merge, on = 'user_id')

In [55]:
df_ords_prods_cust_merged.head()

user_id  gender     state  age date_joined  number_of_dependants  \
0    26711  Female  Missouri   48    1/1/2017                     3   
1    26711  Female  Missouri   48    1/1/2017                     3   
2    26711  Female  Missouri   48    1/1/2017                     3   
3    26711  Female  Missouri   48    1/1/2017                     3   
4    26711  Female  Missouri   48    1/1/2017                     3   

  family_status  income  product_id product_name  ...    price_range_loc  \
0       married  165665         196         Soda  ...  Mid-range product   
1       married  165665         196         Soda  ...  Mid-range product   
2       married  165665         196         Soda  ...  Mid-range product   
3       married  165665        6184  Clementines  ...  Low-range product   
4       married  165665        6184  Clementines  ...  Low-range product   

      busiest_day    busiest_days  busiest_period_of_day  max_order  \
0  Regularly busy    Busiest days            Most orders          8   
1  Regularly busy  Regularly busy            Most orders          8   
2  Regularly busy    Busiest days            Most orders          8   
3  Regularly busy  Regularly busy            Most orders          8   
4  Regularly busy    Slowest days            Most orders          8   

   loyalty_flag  average_spending  spender_flag  purchase_frequency  \
0  New customer          7.988889   Low spender                19.0   
1  New customer          7.988889   Low spender                19.0   
2  New customer          7.988889   Low spender                19.0   
3  New customer          7.988889   Low spender                19.0   
4  New customer          7.988889   Low spender                19.0   

     frequency_flag  
0  Regular customer  
1  Regular customer  
2  Regular customer  
3  Regular customer  
4  Regular customer  

[5 rows x 32 columns]

In [56]:
df_ords_prods_cust_merged.shape

(32404859, 32)

Row number of the merged dataframe is equal to the row number of the df_prds_prods_merge dataframe.

# Step 7. Exporting Data

In [ ]:
df_ords_prods_cust_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers_merged.pkl'))